tests/end-to-end/upgrade/pkg/tests/service-catalog/flow.go

```diff
@@ -0,0 +1,310 @@
+		Spec: v1beta1.ServiceBindingSpec{
+			InstanceRef: v1beta1.LocalObjectReference{
+				Name: instanceName,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	f.log.Infof("Waiting for binding %s to be ready", bindingName)
+	return f.wait(60*time.Second, func() (done bool, err error) {
+		sb, err := f.scInterface.ServicecatalogV1beta1().ServiceBindings(f.namespace).Get(bindingName, metav1.GetOptions{})
+		if err != nil {
+			f.log.Error(err.Error())
+			return false, err
+		}
+		for _, cnd := range sb.Status.Conditions {
+			if cnd.Type == v1beta1.ServiceBindingConditionReady && cnd.Status == v1beta1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+}
+
+func (f *flow) createBindingUsageAndWaitForReadiness(usageName, bindingName, deploymentName string) error {
+	f.log.Infof("Creating %s binding usage", usageName)
+	_, err := f.buInterface.ServicecatalogV1alpha1().ServiceBindingUsages(f.namespace).Create(&v1alpha1.ServiceBindingUsage{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ServiceBindingUsage",
+			APIVersion: "servicecatalog.kyma-project.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: usageName,
+		},
+		Spec: v1alpha1.ServiceBindingUsageSpec{
+			ServiceBindingRef: v1alpha1.LocalReferenceByName{
+				Name: bindingName,
+			},
+			UsedBy: v1alpha1.LocalReferenceByKindAndName{
+				Kind: "deployment",
+				Name: deploymentName,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	f.log.Infof("Waiting for binding usage %s to be ready", usageName)
+	return f.wait(5*time.Second, func() (done bool, err error) {
+		si, err := f.buInterface.ServicecatalogV1alpha1().ServiceBindingUsages(f.namespace).Get(usageName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, cnd := range si.Status.Conditions {
+			if cnd.Type == v1alpha1.ServiceBindingUsageReady && cnd.Status == v1alpha1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+}
+
+func (f *flow) deleteServiceBinding(name string) error {
+	f.log.Infof("Removing Service Binding %s", name)
+	return f.scInterface.ServicecatalogV1beta1().ServiceBindings(f.namespace).Delete(name, &metav1.DeleteOptions{})
+}
+
+func (f *flow) deleteServiceInstance(name string) error {
+	f.log.Infof("Removing Service Instance %s", name)
+	return f.scInterface.ServicecatalogV1beta1().ServiceInstances(f.namespace).Delete(name, &metav1.DeleteOptions{})
+}
+
+func (f *flow) deleteBindingUsage(name string) error {
+	f.log.Infof("Removing Service Binding Usage %s", name)
+	return f.buInterface.ServicecatalogV1alpha1().ServiceBindingUsages(f.namespace).Delete(name, &metav1.DeleteOptions{})
+}
+
+func (f *flow) wait(timeout time.Duration, conditionFunc wait.ConditionFunc) error {
+	timeoutCh := time.After(timeout)
+	stopCh := make(chan struct{})
+	go func() {
+		select {
+		case <-timeoutCh:
+			close(stopCh)
+		case <-f.stop:
+			close(stopCh)
+		}
+	}()
+	return wait.PollUntil(500*time.Millisecond, conditionFunc, stopCh)
```

Comment:
I do not know if it's a good approach to use that. Under the hood it uses that function:
https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L398-L419

and when timeout occurs u will get only information:
`timed out waiting for the condition`

but will not get the last error, so it's hard to determine why test failed

Reply:
If the conditionFunc returns an error, you will see that. Let's discuss